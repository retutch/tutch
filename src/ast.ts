export interface Syn {
  readonly type: string;
  readonly range?: [number, number];
  readonly loc?: SourceLocation;
}

export interface Position {
  readonly line: number;
  readonly column: number;
}

export interface SourceLocation {
  readonly start: Position;
  readonly end: Position;
  readonly source: string | null;
}

export type Term = TermVar | TermConst;

export interface TermVar extends Syn {
  readonly type: 'TermVar';
  readonly index: number;
}

export interface TermConst extends Syn {
  readonly type: 'TermConst';
  readonly head: string;
  readonly spine: Term[];
}

export type Proposition =
  | PropTrue
  | PropFalse
  | Atom
  | PropAnd
  | PropImplies
  | PropOr
  | PropAll
  | PropExists;

export interface PropTrue extends Syn {
  readonly type: 'PropTrue';
}

export interface PropFalse extends Syn {
  readonly type: 'PropFalse';
}

export interface Atom extends Syn {
  readonly type: 'Atom';
  readonly predicate: string;
  readonly spine: Term[];
}

export interface PropAnd extends Syn {
  readonly type: 'PropAnd';
  readonly left: Proposition;
  readonly right: Proposition;
}

export interface PropImplies extends Syn {
  readonly type: 'PropImplies';
  readonly left: Proposition;
  readonly right: Proposition;
}

export interface PropOr extends Syn {
  readonly type: 'PropOr';
  readonly left: Proposition;
  readonly right: Proposition;
}

export interface PropAll extends Syn {
  readonly type: 'PropAll';
  readonly variable: string;
  readonly sort: 't' | 'nat';
  readonly argument: Proposition;
}

export interface PropExists extends Syn {
  readonly type: 'PropExists';
  readonly variable: string;
  readonly sort: 't' | 'nat';
  readonly argument: Proposition;
}

function freshenRelativeTo(sigma: string[], x: string): string {
  let i = 0;
  let z = x;
  while (sigma.some((y) => z === y)) {
    i += 1;
    z = `${x}${i}`;
  }

  return z;
}

export function termToString(sigma: string[], term: Term): string {
  switch (term.type) {
    case 'TermConst': {
      if (term.spine.length === 0) {
        return term.head;
      } else {
        return `(${term.head}${term.spine.map((tm) => ` ${termToString(sigma, tm)}`).join('')})`;
      }
    }
    case 'TermVar': {
      return sigma[term.index];
    }
  }
}

export function propToString(sigma: string[], prop: Proposition): string {
  switch (prop.type) {
    case 'PropOr':
      return `(${propToString(sigma, prop.left)} | ${propToString(sigma, prop.right)})`;
    case 'PropAnd':
      return `(${propToString(sigma, prop.left)} & ${propToString(sigma, prop.right)})`;
    case 'PropImplies':
      return `(${propToString(sigma, prop.left)} => ${propToString(sigma, prop.right)})`;
    case 'PropTrue':
      return 'T';
    case 'PropFalse':
      return 'F';
    case 'PropAll': {
      const x = freshenRelativeTo(sigma, prop.variable);
      return `(!${x}:${prop.sort}.${propToString([x].concat(sigma), prop.argument)})`;
    }
    case 'PropExists': {
      const x = freshenRelativeTo(sigma, prop.variable);
      return `(?${x}:${prop.sort}.${propToString([x].concat(sigma), prop.argument)})`;
    }
    case 'Atom':
      return `${prop.predicate}${prop.spine.map((tm) => ` ${termToString(sigma, tm)}`).join('')}`;
  }
}

export function termToStringDebug(term: Term): string {
  switch (term.type) {
    case 'TermVar': {
      return `#${term.index}`;
    }
    case 'TermConst': {
      if (term.spine.length === 0) {
        return term.head;
      } else {
        return `(${term.head}${term.spine.map((tm) => ` ${termToStringDebug(tm)}`).join('')})`;
      }
    }
  }
}

export function propToStringDebug(prop: Proposition): string {
  switch (prop.type) {
    case 'PropOr':
      return `(${propToStringDebug(prop.left)} | ${propToStringDebug(prop.right)})`;
    case 'PropAnd':
      return `(${propToStringDebug(prop.left)} & ${propToStringDebug(prop.right)})`;
    case 'PropImplies':
      return `(${propToStringDebug(prop.left)} => ${propToStringDebug(prop.right)})`;
    case 'PropTrue':
      return 'T';
    case 'PropFalse':
      return 'F';
    case 'PropAll': {
      return `(!${prop.variable}:${prop.sort}.${propToStringDebug(prop.argument)})`;
    }
    case 'PropExists': {
      return `(?${prop.variable}:${prop.sort}.${propToStringDebug(prop.argument)})`;
    }
    case 'Atom':
      return `${prop.predicate}${prop.spine.map((tm) => ` ${termToStringDebug(tm)}`).join('')}`;
  }
}

export type ProofStep = Proposition | HypotheticalProof;

export interface VariableDeclaration extends Syn {
  readonly type: 'VariableDeclaration';
  readonly variable: string;
  readonly sort: 't';
}

export type Hypothesis = Proposition | VariableDeclaration;

export interface HypotheticalProof extends Syn {
  readonly type: 'HypotheticalProof';
  hypotheses: Hypothesis[];
  steps: ProofStep[];
  consequent: Proposition;
}

export interface Proof extends Syn {
  readonly type: 'Proof';
  readonly name: string;
  readonly goal: Proposition;
  readonly proof: ProofStep[];
  readonly consequent: Proposition; // Must be same as goal
}

export function equalTerms(a: Term, b: Term): boolean {
  switch (a.type) {
    case 'TermVar': {
      return a.type === b.type && a.index === b.index;
    }
    case 'TermConst': {
      return (
        a.type === b.type &&
        a.head === b.head &&
        a.spine.length === b.spine.length &&
        a.spine.every((tm, i) => equalTerms(tm, b.spine[i]))
      );
    }
  }
}

export function equalProps(a: Proposition, b: Proposition): boolean {
  switch (a.type) {
    case 'Atom':
      return (
        a.type === b.type &&
        a.predicate === b.predicate &&
        a.spine.length === b.spine.length &&
        a.spine.every((tm, i) => equalTerms(tm, b.spine[i]))
      );
    case 'PropTrue':
    case 'PropFalse':
      return a.type === b.type;
    case 'PropAnd':
    case 'PropImplies':
    case 'PropOr':
      return a.type === b.type && equalProps(a.left, b.left) && equalProps(a.right, b.right);
    case 'PropAll':
      return a.type === b.type && a.sort === b.sort && equalProps(a.argument, b.argument);
    case 'PropExists':
      return a.type === b.type && a.sort === b.sort && equalProps(a.argument, b.argument);
  }
}
