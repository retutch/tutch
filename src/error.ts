import { SourceLocation } from './ast';

export class ParsingError extends Error {
  public readonly name: 'ParsingError' = 'ParsingError';
  loc: null | SourceLocation;
  constructor(syn: SourceLocation | { loc?: SourceLocation }, msg: string) {
    super(msg);
    const loc = 'start' in syn ? syn : syn.loc ? syn.loc : null;
    this.loc = loc;
  }
}
