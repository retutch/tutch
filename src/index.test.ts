import { readdirSync, readFileSync, lstatSync } from 'fs';
import { join, extname } from 'path';
import * as Ast from './ast';
import { evaluateAssert, isJustified, parse, parseProp } from '.';
import { Parser, Grammar } from 'nearley';
import { ImpossibleError } from './error';
//import { Proof } from './ast';

const testSpecRules = require('../dist/spec-rules');

function itShouldRejectWithParsingError(str: string) {
    test(`cannot parse '${str}' as a proposition`, async () => {
        expect(() => parseProp(str)).toThrowErrorMatchingSnapshot();
    });
}

function itShouldParseAndPrint(input: string, output: string) {
    test(`parsing and printing '${input}' produces '${output}'`, async () => {
        expect(Ast.propToStringDebug(parseProp(input))).toEqual(output);
    });
}

describe('proposition parser', () => {
    itShouldRejectWithParsingError('');
    itShouldRejectWithParsingError('(');
    itShouldRejectWithParsingError(')');
    itShouldRejectWithParsingError('â˜„');
    itShouldRejectWithParsingError('lower & Upper');
    itShouldRejectWithParsingError('Upper & lower');
    itShouldRejectWithParsingError('!x:custom.A');
    itShouldRejectWithParsingError('nat => A');
    itShouldRejectWithParsingError('A & let');
    itShouldRejectWithParsingError('!A:t. B');
    itShouldRejectWithParsingError('?A:t. B');
    itShouldRejectWithParsingError('A x y & B y C');
    itShouldRejectWithParsingError('!h. A (h x)');
    itShouldRejectWithParsingError('A B');

    itShouldParseAndPrint('!x. A(h x)', '(!x:t.A (h #0))');
    itShouldParseAndPrint('A x y z', 'A x y z');
    itShouldParseAndPrint('A => B', '(A => B)');
    itShouldParseAndPrint('A => B => C', '(A => (B => C))');
    itShouldParseAndPrint('A & B => C & D', '((A & B) => (C & D))');
    itShouldParseAndPrint('A & !a:t. B & C', '(A & (!a:t.(B & C)))');
    itShouldParseAndPrint('A => !a:t. B & C', '(A => (!a:t.(B & C)))');
    itShouldParseAndPrint('A & !a:t. B => C', '(A & (!a:t.(B => C)))');
    itShouldParseAndPrint('A | ?a:t. B => C', '(A | (?a:t.(B => C)))');
    itShouldParseAndPrint('A => ?a:t. B | C', '(A => (?a:t.(B | C)))');
    itShouldParseAndPrint('F => ?a:t. B a | C b', '(F => (?a:t.(B #0 | C b)))');
    itShouldParseAndPrint('T => !b:t. B a | C b', '(T => (!b:t.(B a | C #0)))');
    itShouldParseAndPrint('!a. ?b. A a b c', '(!a:t.(?b:t.A #1 #0 c))');
    itShouldParseAndPrint('?x. A x & ?x. A x', '(?x:t.(A #0 & (?x:t.A #0)))');
    itShouldParseAndPrint('!x. A x & !x. A x', '(!x:t.(A #0 & (!x:t.A #0)))');
});

describe('evaluateAssert', () => {
    test('succeeds on a valid proof', async () => {
        expect(() => evaluateAssert(parse('proof x : T = begin T end;'))).not.toThrow();
    });
    test('throws on an invalid proof', async () => {
        expect(() => evaluateAssert(parse('proof x : F = begin F end;'))).toThrow();
    });
});


const dir = './tests';

export interface Spec {
    outcome: 'error' | 'reject' | 'pass';
    description: string;
}

export function parseSpec(spec: string, filename?: string): Spec[] {
    let specs;
    try {
        const specParser = new Parser(Grammar.fromCompiled(testSpecRules));
        specParser.feed(spec);
        specs = specParser.finish();
    } catch (err) {
        throw new Error(`Error parsing test spec ${filename ? `for ${filename}` : ''}:\n${err}`);
    }

    if (specs.length === 0) throw new Error('No test spec found');
    /* istanbul ignore next */
    if (specs.length > 1) {
        throw new ImpossibleError('Test spec parsing ambiguous');
    }

    return specs[0][0].map((spec: any) => {
        return ((cond: string): Spec => {
            switch (cond) {
                case 'error':
                    return { outcome: 'error', description: 'be flagged as an ill-formed file' };
                case 'reject':
                    return { outcome: 'reject', description: 'be rejected' };
                case 'pass':
                    return { outcome: 'pass', description: 'pass' };
                /* istanbul ignore next */
                default:
                    throw new Error(`Unexpected condition ${cond} (should be impossible, please report)`);
            }
        })(spec[2][0]);
    });
}

function testfile(filepath: string) {
    const contents = readFileSync(filepath, { encoding: 'binary' });
    const spectxt = contents.match(/(% test .*(\r|\n|\n\r|\r\n))+/);
    if (spectxt === null) {
        console.warn(`No specs in file ${filepath}`);
        return;
    }

    let specs: Spec[];
    try {
        specs = parseSpec(spectxt[0], filepath);
    } catch (err: any) {
        console.log(err.message);
        specs = [];
    }

    specs.forEach((spec, i) => {
        test(`${filepath}${i ? `.${i}` : ''} should ${spec.description}`, async () => {
            console.log(spec.outcome);
            switch (spec.outcome) {
                case 'error':
                    expect(() => parse(contents)).toThrowErrorMatchingSnapshot();
                    break;
                case 'reject':
                    expect(isJustified(parse(contents))).toEqual(false);
                    break;
                default:
                    expect(isJustified(parse(contents))).toEqual(true);
                    break;
            }
        });
    });
}

readdirSync(dir).forEach((subdir) => {
    if (lstatSync(join(dir, subdir)).isDirectory()) {
        describe(`Tests in suite ${subdir}`, () => {
            readdirSync(join(dir, subdir)).forEach((file) => {
                const ext = extname(file);
                if (ext === '.tut') {
                    testfile(join(dir, subdir, file));
                }
            });
        });
    }
});
