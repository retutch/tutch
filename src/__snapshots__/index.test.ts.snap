// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in suite firstorder tests/firstorder/non-t-sort.tut should be flagged as an ill-formed file 1`] = `"Only \`t\` is allowed as a first-order type, not custom"`;

exports[`Tests in suite prop tests/prop/hyphyp.tut should be flagged as an ill-formed file 1`] = `"Hypothetical proof [...] must end with a proposition, not another hypothetical sequence."`;

exports[`Tests in suite prop tests/prop/lost-track.tut should be flagged as an ill-formed file 1`] = `"The last line in a proof must be the proof's goal proposition."`;

exports[`Tests in suite prop tests/prop/mp-error.tut should be flagged as an ill-formed file 1`] = `"Unexpected input ["`;

exports[`Tests in suite prop tests/prop/noproof.tut should be flagged as an ill-formed file 1`] = `"Unexpected input ]"`;

exports[`Tests in suite prop tests/prop/nowork.tut should be flagged as an ill-formed file 1`] = `"Unexpected input ]"`;

exports[`proposition parser cannot parse '!A:t. B' as a proposition 1`] = `"Term variables must start with a lower case letter"`;

exports[`proposition parser cannot parse '!h. A (h x)' as a proposition 1`] = `"Bound variable 'h' cannot have arguments applied to it."`;

exports[`proposition parser cannot parse '!x:custom.A' as a proposition 1`] = `"Only \`t\` is allowed as a first-order type, not custom"`;

exports[`proposition parser cannot parse '' as a proposition 1`] = `"Incomplete parse at the end of the file"`;

exports[`proposition parser cannot parse '(' as a proposition 1`] = `"Incomplete parse at the end of the file"`;

exports[`proposition parser cannot parse ')' as a proposition 1`] = `"Unexpected input )"`;

exports[`proposition parser cannot parse '?A:t. B' as a proposition 1`] = `"Term variables must start with a lower case letter"`;

exports[`proposition parser cannot parse 'A & let' as a proposition 1`] = `"Predicate names must start with an upper case letter"`;

exports[`proposition parser cannot parse 'A B' as a proposition 1`] = `"Term variables and constants must start with a lower case letter"`;

exports[`proposition parser cannot parse 'A x y & B y C' as a proposition 1`] = `"Term variables and constants must start with a lower case letter"`;

exports[`proposition parser cannot parse 'Upper & lower' as a proposition 1`] = `"Predicate names must start with an upper case letter"`;

exports[`proposition parser cannot parse 'lower & Upper' as a proposition 1`] = `"Predicate names must start with an upper case letter"`;

exports[`proposition parser cannot parse 'nat => A' as a proposition 1`] = `"Unexpected input nat"`;

exports[`proposition parser cannot parse '☄' as a proposition 1`] = `"Unexpected input ☄"`;
